### PR TITLE
 Add device OS update action

### PR DESCRIPTION
**Pull request by thgreasi**
That's on top of https://github.com/balena-io/balena-cli/pull/1278 in order to ease reviewing.

Resolves: #788
Depends-on: balena-io/balena-sdk#638
Change-type: minor
Signed-off-by: Thodoris Greasidis <thodoris@balena.io>

---
##### Contributor checklist

- [ ] Introduces security considerations
- [ ] Affects the development, build or deployment processes of the component

### CODE REVIEW

**Comment by thgreasi** [comment]
Just to warn you, this is a Hack Friday proof of concept for adding a way to HUP a device from the CLI.
I already have some things in mind that need to change for this & expect that some of the code of this PR will be abstracted into a newer version of the SDK (eg: the device os update progress helper).

File: lib/actions/device.coffee [hunk 1]
```diff
@@ -455,3 +455,55 @@ exports.init =
 				return device.uuid
 
 		.nodeify(done)
+
+exports.osUpdate =
```
Comment (pdcastro) [inline lib/actions/device.coffee]: Oh no, 40 new lines of coffeescript! :-) Could we create a parallel `device_ts.ts` file, living alongside `device.coffee`, and add just this new `exports.osUpdate` to that ts file? I've done this kind of thing to other files...

Comment (thgreasi) [reply lib/actions/device.coffee]: Yea sure :+1: 
I will wait for the respective SDK parts to get released and re-iterate on this a bit later (eg next Friday).

File: package.json [hunk 2]
```diff
@@ -112,9 +112,11 @@
     "archiver": "^2.1.0",
     "balena-config-json": "^2.0.0",
     "balena-device-init": "^5.0.0",
+    "balena-device-status": "^3.1.2",
     "balena-image-manager": "^6.0.0",
     "balena-preload": "^8.2.0",
     "balena-sdk": "^11.17.0",
+    "balena-semver": "2.0.2",
```
Comment (thgreasi) [inline package.json]: I'm using a pinned version b/c later versions require a typescript upgrade to work, which I tried but ended up complicating this PR a lot.

**Comment by hedss** [comment]
@thgreasi You're gonna hate me for this... we're switching to Oclif so any 'new' commands should really be in this format.

**Comment by thgreasi** [comment]
@hedss I was kind of expecting this to happen eventually... This PR was on hold since Mar 15, so... I guess I'm going to change this on my next Hack Friday.

**Comment by pdcastro** [comment]
> (at)thgreasi You're gonna hate me for this... we're switching to Oclif so any 'new' commands should really be in this format.

> (at)hedss I was kind of expecting this to happen eventually... This PR was on hold since Mar 15, so... I guess I'm going to change this on my next Hack Friday.

@thgreasi, the only command transitioned to oclif in the v11-meta-* branches is `balena env add`, as a "proof of concept" and to prepare the foundations for further transition. So if you have already implemented a new command using Capitano, from my perspective it's OK, no need to modify it right away, it can be modified over time alongside all other Capitano commands -- especially as you've implemented it before v11 was released.

**Comment by thgreasi** [comment]
> no need to modify it right away

Cool @pdcastro :+1: 
In that case I will wait for v11 to get released and for https://github.com/balena-io/balena-cli/pull/1278 to get merged. Give me a note if you change your mind and demand this on oclif.

**Comment by imrehg** [comment]
@thgreasi I was trying this out, but something is off, as the command help, for example, cannot be found. The command is there:

```
?788-hup ~/resin/balena-cli> bin/balena help --verbose          
Usage: balena [COMMAND] [OPTIONS]
[snip]
    device move <uuid>                    move a device to another application                             
    device os-update <uuid>               Start a Host OS update for a device                              
    device public-url <uuid>              gets the public URL of a device                                  
[snip]
```
but the help misunderstands things:
```
?788-hup ~/resin/balena-cli> bin/balena device os-update --help
Usage: device <uuid>

Use this command to show information about a single device.

Examples:

	$ balena device 7cf02a6
```
(it works just fine with `balena device init --help` for example). Am I doing something wrong, or something's missing?

**Comment by thgreasi** [comment]
I did see that @imrehg but it seems like a capitano issue.

**Comment by thgreasi** [comment]
Found it @imrehg , and just pushed a fix!
Thanks :)

**Comment by imrehg** [comment]
@thgreasi looks great! I like the replication of the progress bar in the console too :)

One thing, that the 100% might not be received, when I run an update, after the 90%, the next step just goes on, I wonder if it will be somewhat confusing to people ("why did it end on 90%?"), but it shouldn't be a big deal this time.

```
?788-hup ~/resin/balena-cli> bin/balena device os-update 3341181
? Target OS version 2.36.0+rev2.prod (recommended)
? Host OS updates require a device restart when they complete. Are you sure you want to proceed? Yes
Updating the OS of autumn-moon to v2.36.0+rev2.prod [======================  ] 90%
The device autumn-moon has been updated to v2.36.0+rev2.prod and will restart shortly!
```

File: lib/utils/patterns.ts [hunk 3]
```diff
@@ -253,6 +253,53 @@ export function awaitDevice(uuid: string) {
 	});
 }
 
+export async function awaitDeviceOsUpdate(
```
Comment (pdcastro) [inline lib/utils/patterns.ts]: Just an observation: this `async` function does not have any `await` clauses...

Comment (thgreasi) [reply lib/utils/patterns.ts]: Nice spot, I will remove it and squash all commits.

**Pull request merged**